File: water.java
import java.io.*;
import java.util.*;

public class water
{
/***********************************
 *   MAIN
 ***********************************/
    //Throw the IOException because it is unimportant and only an issue
    //if I handle the input incorrectly or the file doesn't exist
    public static void main(String[] args) throws IOException
    {
    	//create a new instance of the problem solving class
    	water w = new water();
    	
    	//Call the problem solving function of the class
    	w.run();
    }


/***********************************
 *   Class Variables
 ***********************************/

	//setup the needed class wide variables
	private BufferedReader in;
	
	//use these strings to make it easier to change if we had needed to
	private String odd = "WS";
	private String even = "RU";
	

/***********************************
 *   Class Functions
 ***********************************/
	//Constructor for the class
	//Throw the IOException the function creating this
	//so it doesn't clutter code here
	public water() throws IOException
	{
		in = new BufferedReader(new FileReader("water.in"));
	}
	
	
	public void run() throws IOException
	{
		/* Rules of the problem
		 *  odd numbered houses can water on Wednesdays and Saturdays (WS)
		 *  even numbered houses can water on Thursdays and Sundays (RU)
		 */
		 
		/* Special Cases
		 *  100Boo (Number and Address with no space separation)
		 *  What about street names that start with numbers (if no separation of spaces)
		 */
		 
		/* Input Specs
		 * Case count
		 * for each case count
		 *   1 line that contains the number first and then the street
		 *   1 line that contains the watering days
		 */ 
		 
		 //Get the number of Cases and change that to an integer
		 int cases = Integer.parseInt(in.readLine());
		 
		 //Loop through the number of cases and handle it
		 for(int i = 0; i < cases; i++)
		 {
		 	handleCase();
		 }
	}
	
	public void handleCase() throws IOException
	{
		//Read the street address and parse it
		//Under the assumption that there will be a space between the address number and street name
		//Break up string by spaces
		StringTokenizer st = new StringTokenizer(in.readLine()," ");
		
		//The first token is supposed to be a number
		//And is specified as a positive integer so we don't have to worry about
		//really really big numbers that can't be held by integers
		int houseNumber = Integer.parseInt(st.nextToken());
		
		//Find out if the house number is even or not.
		//If it were even then the remainder would not be 0 
		boolean evenNumber = (houseNumber % 2) == 0; 
		
		//Note that the rest of the tokens don't matter
		
		//Grab the days the person watered the lawn.
		String watered = in.readLine();
		
		String badDays = "";
		//Loop through the days this person watered
		//and check to see if they watered on the right day
		for(int i = 0; i < watered.length(); i++)
		{
			//Get the current day
			char theDay = watered.charAt(i);
			
			//if the house number was even and was watered on a day other than those specified
			//add it to the badDays list
			if(evenNumber && even.indexOf(theDay) == -1)
			{
				badDays += theDay;
			}
			//If the house number was odd and was watered on a day other than those specified
			//add it to the badDays list
			else if(!evenNumber && odd.indexOf(theDay) == -1)
			{
				badDays += theDay;
			}
		}
		
		//if badDays length is 0 then they watered on correct days only
		//so print out All safe!  otherwise print out the badDays string
		if(badDays.length() == 0)
		{
			System.out.println("All safe!");
		}
		else
		{
			System.out.println(badDays);
		}
	}
}

File: water.c
#include <stdlib.h>
#include <stdio.h>

#define MAX_S 256

void printInvalidDays(char daysWatered[MAX_S], char* validDays);

int main(){

	FILE* fin;
	char streetAddress[MAX_S];
	char daysWatered[MAX_S];
	int i, n, streetNumber;
	char* sundayThursday = "UR";
	char* wednesdaySaturday = "WS";
	
	/* open the water.in input file for reading
	 **/
	fin = fopen("water.in", "r");
	
	/* read in the number of input cases
	 * Notice the \n after the %d.  This is used so the newline
	 *   character is skipped over.
	 *
	 * Try removing the \n and see how it behaves differently.
	 **/
	fscanf(fin, "%d\n", &n);
	
	/* Loop through all of the input cases
	 * Notice how we never have to check for EOF
	 **/
	for(i = 0; i < n; i++){
		
		/* Reads in the street address line from the file
		 * Here, fgets is used because we have to read in the whole line
		 * We do not know ahead of time how many space seperated words
		 *   will be in this string
		 **/
		fgets(streetAddress, MAX_S, fin);
		
		/* Reads in the days watered line.  Agian, notice the \n
		 **/
		fscanf(fin, "%s\n", daysWatered);
		
		/* Grab the number off of the street address string.
		 * We don't care about the rest of the street address.
		 **/
		sscanf(streetAddress, "%d", &streetNumber);
		
		/* Determine whether or not the street number is odd or even,
		 *   and call the printInvalidDays function with the correct
		 *   parameters.
		 **/
		if(streetNumber % 2 == 0){
			printInvalidDays(daysWatered, sundayThursday);
		}else{
			printInvalidDays(daysWatered, wednesdaySaturday);
		}
		
	}
	
	return 0;
	
}

/* This function takes as its first parameter the days that the house was watered
 *   and the days that the house should be watered.
 * It will print out any characters in daysWatered that are not in validDays
 **/
void printInvalidDays(char daysWatered[MAX_S], char* validDays){

	char* dayWatered;
	char* validDay;
	int wateredOnOffDay, dayIsOK;
	
	/* We use this as a boolean value to indicate whether or not the 
	 *   house was watered on an invalid day.
	 **/
	wateredOnOffDay = 0;
	
	/* Loop through each one of the days the house was watered.
	 **/
	for(dayWatered = &daysWatered[0]; *dayWatered != '\0'; dayWatered++){

		/* This boolean valud indicates whether or not the current dayWatered
		 *   exists in the validDays string.
		 **/
		dayIsOK = 0;

		/* Loop through the validDays and if the dayWatered is equal to any of
		 *  the valid days, we set dayIsOK to true and break out of the loop.
		 **/
		for(validDay = &validDays[0]; *validDay != '\0'; validDay++){
			if(*dayWatered==*validDay){
				dayIsOK = 1;
				break;
			}
		}
		
		/* If the house was watered on an invalid day, print out the character,
		 *   and set wateredOnOffDay to true.
		 **/
		if(!dayIsOK){
			printf("%c", *dayWatered);
			wateredOnOffDay = 1;
		}
		
	}
	
	/* If the house was never watered on an off day, then we print "All safe!"
	 **/
	if(!wateredOnOffDay){
		printf("All safe!");
	}
	printf("\n");
	
}
